File: rust/rfprop/src/sigserve.cc
#include "sigserve.h"

#include <algorithm>
#include <cmath>

#include "../../../src/common.hh"
#include "rfprop/src/sigserve.rs.h"

extern int init(const char *sdf_path, bool debug);
extern int handle_args(int argc, char *argv[], output &ret_out);
extern int LoadTopoData(double max_lon, double min_lon, double max_lat, double min_lat, struct output *out);
extern double LonDiff(double lon1, double lon2);

namespace sigserve_wrapper {

int init(const char *sdf_path, bool debug) { return ::init(sdf_path, debug); };

Report handle_args(int argc, char *argv[])
{
    Report report;
    output out;

    report.retcode = ::handle_args(argc, argv, out);
    if (report.retcode) {
        return report;
    }

    report.dbm = out.dBm;
    report.loss = out.loss;
    report.field_strength = out.field_strength;

    report.cluttervec.reserve(out.cluttervec.size());
    std::copy(out.cluttervec.begin(), out.cluttervec.end(), std::back_inserter(report.cluttervec));

    report.line_of_sight.reserve(out.line_of_sight.size());
    std::copy(out.line_of_sight.begin(), out.line_of_sight.end(), std::back_inserter(report.line_of_sight));

    report.fresnelvec.reserve(out.fresnelvec.size());
    std::copy(out.fresnelvec.begin(), out.fresnelvec.end(), std::back_inserter(report.fresnelvec));

    report.fresnel60vec.reserve(out.fresnel60vec.size());
    std::copy(out.fresnel60vec.begin(), out.fresnel60vec.end(), std::back_inserter(report.fresnel60vec));

    report.curvaturevec.reserve(out.curvaturevec.size());
    std::copy(out.curvaturevec.begin(), out.curvaturevec.end(), std::back_inserter(report.curvaturevec));

    report.profilevec.reserve(out.profilevec.size());
    std::copy(out.profilevec.begin(), out.profilevec.end(), std::back_inserter(report.profilevec));

    report.distancevec.reserve(out.distancevec.size());
    std::copy(out.distancevec.begin(), out.distancevec.end(), std::back_inserter(report.distancevec));

    report.image_data.reserve(out.imagedata.size());
    std::copy(out.imagedata.begin(), out.imagedata.end(), std::back_inserter(report.image_data));

    return report;
}

TerrainProfile terrain_profile(double tx_lat, double tx_lon, double tx_antenna_alt, double rx_lat, double rx_lon,
                               double rx_antenna_alt, double freq_hz, bool normalize, bool metric)
{
    // The following conversions are confusing, and I'm still not
    // entirely sure why negating longitudes works.
    //
    // But here's what I know: internally, uses normal +/- notation
    // for latitude. However, for longitude, it uses [0,360) to
    // represent longitude.
    //
    // An example mapping using actual input parameters to this function:
    //
    // tx_lat: 68.49 (68.49° N)
    // tx_lon: 179.9 (179.9° E)
    // rx_lat: 68.49 (68.49° N)
    // rx_lon: -179.9 (179.9° W)
    //
    // Translates to the following:
    //
    // _min_lat: 68.000000
    // _max_lat: 68.000000
    // _tx_lon: -180.000000
    // _rx_lon: 179.000000
    // _min_lon: 179.000000
    // _max_lon: -180.000000
    //
    // Which causes the following files to be sources in LoadTopoData:
    //
    // 68:69:179:180.bsdf
    // 68:69:180:181.bsdf
    //
    // And calls the terrain profile with the following sites with the
    // original-but-negated longitudes:
    //
    // tx_site { lat: 68.49, lon: -179.9 }
    // rx_site { lat: 68.49, lon: 179.9 }

    double _min_lat = std::floor(std::min(tx_lat, rx_lat));
    double _max_lat = std::floor(std::max(tx_lat, rx_lat));
    double _tx_lon = std::floor(-tx_lon);
    double _rx_lon = std::floor(-rx_lon);
    double _min_lon = LonDiff(_tx_lon, _rx_lon) < 0.0 ? _tx_lon : _rx_lon;
    double _max_lon = LonDiff(_tx_lon, _rx_lon) < 0.0 ? _rx_lon : _tx_lon;

    printf("_min_lat: %f, _max_lat: %f, _tx_lon: %f, _rx_lon: %f, _min_lon: %f, _max_lon: %f\n", _min_lat, _max_lat, _tx_lon,
           _rx_lon, _min_lon, _max_lon);

    LoadTopoData(_max_lon, _min_lon, _max_lat, _min_lat, nullptr);

    site tx_site;
    tx_site.lat = tx_lat;
    tx_site.lon = -tx_lon;
    tx_site.alt = metric ? tx_antenna_alt * FEET_PER_METER : tx_antenna_alt;

    site rx_site;
    rx_site.lat = rx_lat;
    rx_site.lon = -rx_lon;
    rx_site.alt = metric ? rx_antenna_alt * FEET_PER_METER : rx_antenna_alt;

    Path path(tx_site, rx_site);

    /* Call sigserve's TerrainProfile constructor */
    ::TerrainProfile p2p(tx_site, rx_site, path, freq_hz, normalize, metric);

    TerrainProfile report;

    report.distance.reserve(p2p._distance.size());
    std::copy(p2p._distance.begin(), p2p._distance.end(), std::back_inserter(report.distance));

    report.los.reserve(p2p._los.size());
    std::copy(p2p._los.begin(), p2p._los.end(), std::back_inserter(report.los));

    report.fresnel.reserve(p2p._fresnel.size());
    std::copy(p2p._fresnel.begin(), p2p._fresnel.end(), std::back_inserter(report.fresnel));

    report.fresnel60.reserve(p2p._fresnel60.size());
    std::copy(p2p._fresnel60.begin(), p2p._fresnel60.end(), std::back_inserter(report.fresnel60));

    report.curvature.reserve(p2p._curvature.size());
    std::copy(p2p._curvature.begin(), p2p._curvature.end(), std::back_inserter(report.curvature));

    report.terrain.reserve(p2p._terrain.size());
    std::copy(p2p._terrain.begin(), p2p._terrain.end(), std::back_inserter(report.terrain));

    return report;
}

}  // namespace sigserve_wrapper

File: rust/rfprop/src/sigserve.rs
use crate::error::Error;
use log::debug;
use std::{
    ffi::{c_char, CString},
    path::Path,
    sync::Once,
};

static INITIALIZED: Once = Once::new();

pub fn init(bsdf_dir: &Path, debug: bool) -> Result<(), Error> {
    let mut ret = 0;

    let bsdf_dir = {
        let mut bsdf_dir = bsdf_dir.to_string_lossy().into_owned().into_bytes();

        // The C++ requires a trailing slash, but it's easier to do
        // that check and conversion here.
        if bsdf_dir.last() != Some(&b'/') {
            bsdf_dir.push(b'/');
        }
        // We need to manually add the null terminator.
        bsdf_dir.push(0);
        CString::from_vec_with_nul(bsdf_dir).unwrap()
    };

    // SAFETY: we're calling into a >20 y/o C+ codebase. Safety can
    //         not be guaranteed.
    INITIALIZED.call_once(|| unsafe {
        ret = ffi::init(bsdf_dir.as_ptr(), debug);
    });

    match ret {
        0 => Ok(()),
        other => Err(Error::Retcode(other)),
    }
}

pub fn get_elevation(lat: f64, lon: f64) -> f64 {
    assert!(
        INITIALIZED.is_completed(),
        "must init rfprop with tile path"
    );
    // SAFETY: See previous safety comment.
    unsafe { ffi::get_elevation(lat, lon) }
}

pub fn call_sigserve(args: &str) -> Result<ffi::Report, Error> {
    assert!(
        INITIALIZED.is_completed(),
        "must init rfprop with tile path"
    );
    let cstrings = args
        .split_whitespace()
        .map(CString::new)
        .collect::<Result<Vec<CString>, _>>()?;
    let mut c_args = cstrings
        .iter()
        .map(|arg| arg.as_ptr() as *mut c_char)
        .collect::<Vec<*mut c_char>>();
    let start = std::time::Instant::now();
    // SAFETY: See previous safety comment.
    let report = unsafe { ffi::handle_args(c_args.len() as i32, c_args.as_mut_ptr()) };
    let duration = start.elapsed();
    debug!("call_sigserve took {:?}", duration);
    match report.retcode {
        0 => Ok(report),
        other => Err(Error::Retcode(other)),
    }
}

#[allow(clippy::too_many_arguments)]
pub fn terrain_profile(
    tx_lat: f64,
    tx_lon: f64,
    tx_antenna_alt_m: f64,
    rx_lat: f64,
    rx_lon: f64,
    rx_antenna_alt_m: f64,
    freq_hz: f64,
    normalize: bool,
) -> ffi::TerrainProfile {
    assert!(
        INITIALIZED.is_completed(),
        "must init rfprop with tile path"
    );

    let start = std::time::Instant::now();
    // SAFETY: this returns no error code nor exception
    let ret = unsafe {
        ffi::terrain_profile(
            tx_lat,
            tx_lon,
            tx_antenna_alt_m,
            rx_lat,
            rx_lon,
            rx_antenna_alt_m,
            freq_hz,
            normalize,
            true,
        )
    };
    let duration = start.elapsed();
    debug!("terrain_profile(tx_lat: {tx_lat}, tx_lon: {tx_lon}, tx_antenna_alt_m: {tx_antenna_alt_m}, rx_lat: {rx_lat}, rx_lon: {rx_lon}, rx_antenna_alt_m: {rx_antenna_alt_m}, freq_hz: {freq_hz}, normalize: {normalize}) took {:?}", duration);
    ret
}

#[cxx::bridge(namespace = "sigserve_wrapper")]
pub(crate) mod ffi {
    #[derive(Default, Debug)]
    pub struct Report {
        // common return code
        retcode: i32,
        // begin point-to-point fields.
        dbm: f64,
        loss: f64,
        field_strength: f64,
        distancevec: Vec<f64>,
        cluttervec: Vec<f64>,
        line_of_sight: Vec<f64>,
        fresnelvec: Vec<f64>,
        fresnel60vec: Vec<f64>,
        curvaturevec: Vec<f64>,
        profilevec: Vec<f64>,
        // end point-to-point fields.

        // begin image-mode fields.
        image_data: Vec<u8>,
        // end image-mode fields.
    }

    #[derive(Default, Debug)]
    pub struct TerrainProfile {
        distance: Vec<f64>,
        los: Vec<f64>,
        fresnel: Vec<f64>,
        fresnel60: Vec<f64>,
        curvature: Vec<f64>,
        terrain: Vec<f64>,
        tx_site_over_water: bool,
    }

    unsafe extern "C++" {
        include!("rfprop/src/sigserve.h");

        unsafe fn init(bsdf_dir: *const c_char, debug: bool) -> i32;

        unsafe fn get_elevation(lat: f64, lon: f64) -> f64;

        unsafe fn handle_args(argc: i32, argv: *mut *mut c_char) -> Report;

        #[allow(clippy::too_many_arguments)]
        unsafe fn terrain_profile(
            tx_lat: f64,
            tx_lon: f64,
            tx_antenna_alt: f64,
            rx_lat: f64,
            rx_lon: f64,
            rx_antenna_alt: f64,
            freq_hz: f64,
            normalize: bool,
            metric: bool,
        ) -> TerrainProfile;
    }
}
